common/persistence/sqlMetadataPersistence.go

```diff
@@ -0,0 +1,429 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package persistence
+
+import (
+	"fmt"
+	_ "github.com/go-sql-driver/mysql" // MySQL driver
+	"github.com/jmoiron/sqlx"
+
+	"bytes"
+	"database/sql"
+	"encoding/gob"
+	workflow "github.com/uber/cadence/.gen/go/shared"
+)
+
+type (
+	sqlMetadataManager struct {
+		db *sqlx.DB
+	}
+
+	FlatCommon struct {
+		DomainInfo
+		DomainConfig
+		ActiveClusterName string `db:"active_cluster_name"`
+		Clusters          []byte `db:"clusters"`
+		ConfigVersion     int64  `db:"config_version"`
+		FailoverVersion   int64  `db:"failover_version"`
+	}
+
+	FlatUpdateDomainRequest struct {
+		FlatCommon
+		FailoverNotificationVersion int64 `db:"failover_notification_version"`
+		NotificationVersion         int64 `db:"notification_version"`
+	}
+
+	flatCreateDomainRequest struct {
+		FlatCommon
+		IsGlobalDomain bool `db:"is_global_domain"`
+	}
+
+	flatGetDomainResponse struct {
+		FlatUpdateDomainRequest
+		IsGlobalDomain bool `db:"is_global_domain"`
+	}
+
+	domainRow = flatGetDomainResponse
+)
+
+func (m *sqlMetadataManager) ListDomain(request *ListDomainRequest) (*ListDomainResponse, error) {
+
+	var rows []*domainRow
+
+	m.db.Select(&rows, templateListDomainSqlQuery)
+
+	var domains []*GetDomainResponse
+	for _, row := range rows {
+		resp, err := domainRowToGetDomainResponse(row)
+		if err != nil {
+			return &ListDomainResponse{
+				Domains: domains,
+			}, err
+		}
+		domains = append(domains, resp)
+	}
+
+	return &ListDomainResponse{
+		Domains: domains,
+	}, nil
+}
+
+const (
+	templateCreateDomainSqlQuery = `INSERT INTO domains (
+		id,
+		retention_days, 
+		emit_metric,
+		config_version,
+		name, 
+		status, 
+		description, 
+		owner_email,
+		failover_version, 
+		is_global_domain,
+		active_cluster_name, 
+		clusters, 
+		notification_version,
+		failover_notification_version
+		)
+		VALUES(
+		:id,
+		:retention_days, 
+		:emit_metric,
+		:config_version,
+		:name, 
+		:status, 
+		:description, 
+		:owner_email,
+		:failover_version, 
+		:is_global_domain,
+		:active_cluster_name, 
+		:clusters,
+		:notification_version,
+		:failover_notification_version
+		)`
+
+	getDomainPart = `SELECT
+		id,
+		retention_days, 
+		emit_metric,
+		config_version,
+		name, 
+		status, 
+		description, 
+		owner_email,
+		failover_version, 
+		is_global_domain,
+		active_cluster_name, 
+		clusters,
+		notification_version,
+		failover_notification_version
+FROM domains
+`
+	templateGetDomainByIdSqlQuery = getDomainPart +
+		`WHERE id = :id`
+	templateGetDomainByNameSqlQuery = getDomainPart +
+		`WHERE name = :name`
+
+	templateUpdateDomainSqlQuery = `UPDATE domains
+SET
+		id = :id,
```

Comment:
id should never be changed. it should on the other hand always be in the where clause.